text/0011-peer-chaincode-optimization.md

```diff
@@ -0,0 +1,171 @@
+
+### Objectives of the Optimization
+The proposed batching mechanism aims to address these challenges by:
+- **Reducing Message Volume:** Grouping multiple state changes into a single `ChangeStateBatch` message reduces the number of messages exchanged between chaincode and peers.
+- **Improving Resource Utilization:** By transmitting fewer messages, the network load is reduced, and processing overhead on both the peer and chaincode is minimized.
+- **Ensuring Backward Compatibility:** Existing applications can continue to function as before without modification. If the batching feature is not supported by a component, the system will gracefully fall back to the original communication model.
+
+---
+
+## Problem Illustration with Sample Chaincode
+The following chaincode example demonstrates the inefficiencies caused by the current communication protocol. Each iteration of the loop sends an individual `PutState` message to the peer node, resulting in unnecessary overhead:
+
+```go
+type OnlyPut struct{}
+
+// Init initializes the chaincode.
+func (t *OnlyPut) Init(_ shim.ChaincodeStubInterface) *pb.Response {
+    return shim.Success(nil)
+}
+
+// Invoke - Entry point for invocations.
+func (t *OnlyPut) Invoke(stub shim.ChaincodeStubInterface) *pb.Response {
+    function, args := stub.GetFunctionAndParameters()
+    switch function {
+    case "invoke":
+        return t.put(stub, args)
+    default:
+        return shim.Error("Received unknown function invocation")
+    }
+}
+
+// Both params should be marshaled JSON data and base64 encoded.
+func (t *OnlyPut) put(stub shim.ChaincodeStubInterface, args []string) *pb.Response {
+    if len(args) != 1 {
+        return shim.Error("Incorrect number of arguments. Expecting 1")
+    }
+    num, _ := strconv.Atoi(args[0])
+    for i := 0; i < num; i++ {
+        key := "key" + strconv.Itoa(i)
+        if err := stub.PutState(key, []byte(key)); err != nil {
+            return shim.Error(err.Error())
+        }
+    }
+    return shim.Success(nil)
+}
+```
+
+---
+
+## Proposed Solution
+
+The proposed solution introduces a **batching mechanism** where multiple state changes are grouped into a single `ChangeStateBatch` message. This new message type will encapsulate all state changes and transmit them as a batch to the peer node. During the `Ready` message exchange, the peer and chaincode will negotiate the capability to use batching, ensuring backward compatibility.
+
+### New Message Format
+```proto
+message ChangeStateBatch {
+    repeated StateKV kvs = 1;
+}
+
+message StateKV {
+    enum Type {
+        UNDEFINED = 0;
+        PUT_STATE = 9;
+        DEL_STATE = 10;
+        PUT_STATE_METADATA = 21;
+        PURGE_PRIVATE_DATA = 23;
+    }
+
+    string key = 1;
+    bytes value = 2;
+    string collection = 3;
+    StateMetadata metadata = 4;
+    Type type = 5;
+}
```

Comment:
Please note that this RFC suggests sending anything that changes state in batches.
I think it's cool.